design.md

```diff
@@ -0,0 +1,121 @@
+
+Goal: Create an API to run arbitrary linux processes on a host utilizing a CLI
+High Level Overview
+
+The proposed solution will be divided into three parts:
+1. The server side library written in Go to interact with the Linux OS
+2. The gRPC API to take in requests from the client, use the server side library to run jobs, and return responses to the client
+3. The CLI that the client will use to send requests related to jobs to the API
+
+## Scope
+
+### In Scope
+* Starting a process
+* Terminating a process
+* Getting a process’s status
+* Streaming the output of a process 
+
+### Not in scope
+* Listing jobs
+* Stopping and restarting processes
+* Task resource management (CPU, memory, etc)
+* Automated certificate creation
+
+## Approach
+The current approach for this project is to create something as simple and straightforward as possible. This means that I will not be using any external libraries or dependencies, and will try to leverage the Linux operating system’s existing tooling as much as possible instead of trying to reinvent the wheel. 
+
+This design will cut a lot of corners and leave a lot of ideal functionality out of the initial implementation. I have noted things that should be considered as `Future Improvements` in each section.
+
+## API
+The API will use gRPC to route requests from the client to the server. The API will run locally on the machine that is being managed from the client.
+
+The API will accept requests for running processes as well as requests to return the status and output of a job. More detailed implementation details for the API can be found in the .proto document.
+
+At a high level the API will have the following methods on the localhost using port 9090:
+
+### Run/Stop Methods
+`RunProcess(RunProcessRequest)`
+`StopProcess(StopProcessRequest)`
+
+### Get Information Methods
+`GetProcessStatus(ProcessStatusRequest)`
+`GetProcessOutput(ProcessOutputRequest)`
+
+The `RunProcess` method will return the process identifier for the command issued to the client, which will be used for the `StopProcess`, `GetProcessStatus`, and `GetProcessOutput` methods.
+
+### Future Improvements
+* Run the server in a separate environment, such as kubernetes, which can allow for defining high availability through replicaSets and would move the API out from the domain of the operating system being managed
+* Utilize some kind of data store (relational or not) to track processes instead of relying on files
+* Include resource management of some kind to keep clients from over-taxing the operating system commands are running on
+* Scope processes to users that map back to the clients on the operating system
+
+## Server Authentication
+
+### mTLS
+The service will utilize mTLS to authenticate and secure requests. For this implementation we will manually create a single local CA certificate that will be used to generate a server side certificate and key pair. These certificates and keys will be stored locally for the server.
```

Comment:
This describes how you will generate the server certificate, but that's TLS, not mTLS.

What about the client certificate(s)?